addons/bus/static/src/workers/election_worker.js

```diff
@@ -0,0 +1,120 @@
+import { Deferred } from "@bus/workers/bus_worker_utils";
+
+export class ElectionWorker {
+    MAIN_TAB_TIMEOUT_PERIOD = 3000;
+
+    constructor() {
+        this.masterTab = null;
+        this.candidates = new Set();
+        this.lastHeartbeat = Date.now();
+        this.electionDeferred = new Deferred();
+        this.masterReplyDeferred = null;
+        this.heartbeatCheckInterval = null;
+        this.heartbeatRequestInterval = null;
+    }
+
+    registerCandidate(messagePort) {
+        this.candidates.add(messagePort);
+        if (this.candidates.size === 1) {
+            this.startElection();
+        }
+    }
+
+    unregisterCandidate(messagePort) {
+        this.candidates.delete(messagePort);
+        if (this.masterTab === messagePort) {
+            this.startElection();
+        }
+    }
+
+    broadcast(message) {
+        for (const candidate of this.candidates) {
+            candidate.postMessage(message);
+        }
+    }
+
+    requestHeartbeat(messagePort) {
+        if (messagePort) {
+            messagePort.postMessage({
+                type: "ELECTION:HEARTBEAT_REQUEST",
+            });
+        } else {
+            this.broadcast({
+                type: "ELECTION:HEARTBEAT_REQUEST",
+            });
+        }
+    }
+
+    async checkMaster() {
+        if (!this.masterReplyDeferred) {
+            this.masterReplyDeferred = new Deferred();
+        }
+        if (!this.electionDeferred) {
+            this.requestHeartbeat(this.masterTab);
+        }
+        await this.masterReplyDeferred;
+    }
+
+    startElection() {
+        this.masterTab = null;
+        clearInterval(this.heartbeatCheckInterval);
+        clearInterval(this.heartbeatRequestInterval);
+        if (!this.electionDeferred) {
+            this.electionDeferred = new Deferred();
+        }
+        this.lastHeartbeat = Date.now();
```

Comment:
Why do we set `lastHeartbeat`? I interpret it as `lastMasterHeartbeat`, but here we set it from the shared worker?